otherlibs/stdune/src/compact_position.ml

```diff
@@ -0,0 +1,134 @@
+module Position = struct
+  (* We encode the position in three, 21 bit fields: [cnum][lnum][bol] *)
+  type t = int
+
+  let field_size = 21
+  let mask_bol = (1 lsl field_size) - 1
+  let mask_lnum = mask_bol lsl field_size
+  let mask_cnum = mask_lnum lsl field_size
+
+  let small_enough =
+    let max_size = 1 lsl field_size in
+    let test int = int <= max_size in
+    fun [@inline] { Lexing.pos_bol; pos_cnum; pos_lnum; pos_fname = _ } ->
+      test pos_bol && test pos_cnum && test pos_lnum
+  ;;
+
+  let[@inline] of_position { Lexing.pos_bol; pos_cnum; pos_lnum; pos_fname = _ } =
+    let t = pos_cnum land mask_bol in
+    let t = (t lsl field_size) lor (pos_lnum land mask_bol) in
+    (t lsl field_size) lor (pos_bol land mask_bol)
+  ;;
+
+  let[@inline] bol t = t land mask_bol
+  let[@inline] lnum t = (t land mask_lnum) lsr field_size
+  let[@inline] cnum t = (t land mask_cnum) lsr (field_size * 2)
+
+  let to_position t ~fname:pos_fname =
+    let pos_bol = bol t in
+    let pos_cnum = cnum t in
+    let pos_lnum = lnum t in
+    { Lexing.pos_bol; pos_cnum; pos_lnum; pos_fname }
+  ;;
+end
+
+module Same_line_loc = struct
+  (* we encode the location in four, 15 bit chunks
+     [bol][lnum][start_cnum][stop_cnum]
+
+     Note that this leaves us with 3 spare bits. We should probably use them to
+     expand [bol] and [lnum] a little. *)
+  type t = int
+
+  let field_size = 15
+  let mask_bol = (1 lsl field_size) - 1 (* 15 bits set *)
+  let mask_lnum = mask_bol lsl field_size
+  let mask_start_cnum = mask_lnum lsl field_size
+  let mask_stop_cnum = mask_start_cnum lsl field_size
```

Comment:
What if, instead of storing `stop_cnum`, we stored `stop_cnum - start_cnum`? That would likely require fewer bits, which would let us allocate more bits to the other field.

Reply:
Left a CR-someday for this. It sounds like a good change, but I'm not sure how much impact it will have in practice. We already cover the vast majority of locations with this encoding.